// @nteract/notebook-preview/lib/latex.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _mathjaxElectron = require("mathjax-electron");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var MathJax = void 0;
if (global.hasOwnProperty("window")) {
  MathJax = window.MathJax;
}

function isMathJaxOkYet() {
  return window && !window.disableMathJax && typeof MathJax !== "undefined" && window.MathJax && window.MathJax.Hub.Queue;
}

var LatexRenderer = function (_React$PureComponent) {
  _inherits(LatexRenderer, _React$PureComponent);

  function LatexRenderer() {
    _classCallCheck(this, LatexRenderer);

    return _possibleConstructorReturn(this, (LatexRenderer.__proto__ || Object.getPrototypeOf(LatexRenderer)).apply(this, arguments));
  }

  _createClass(LatexRenderer, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      if (isMathJaxOkYet()) (0, _mathjaxElectron.typesetMath)(this.rendered);
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      if (isMathJaxOkYet()) (0, _mathjaxElectron.typesetMath)(this.rendered);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      return _react2.default.createElement(
        "div",
        {
          ref: function ref(rendered) {
            _this2.rendered = rendered;
          }
        },
        this.props.children
      );
    }
  }]);

  return LatexRenderer;
}(_react2.default.PureComponent);

exports.default = LatexRenderer;
