// @nteract/editor/lib/tooltip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tooltipRequest = undefined;
exports.tooltipObservable = tooltipObservable;
exports.tool = tool;

var _Observable = require("rxjs/Observable");

require("rxjs/add/operator/pluck");

require("rxjs/add/operator/first");

require("rxjs/add/operator/map");

var _messaging = require("@nteract/messaging");

function tooltipObservable(channels, editor, message) {
  var tip$ = channels.shell.childOf(message).ofMessageType(["inspect_reply"]).pluck("content").first().map(function (results) {
    return {
      dict: results.data
    };
  });
  // On subscription, send the message
  return _Observable.Observable.create(function (observer) {
    var subscription = tip$.subscribe(observer);
    channels.shell.next(message);
    return subscription;
  });
}

var tooltipRequest = exports.tooltipRequest = function tooltipRequest(code, cursorPos) {
  return (0, _messaging.createMessage)("inspect_request", {
    content: {
      code: code,
      cursor_pos: cursorPos,
      detail_level: 0
    }
  });
};

function tool(channels, editor) {
  var cursor = editor.getCursor();
  var cursorPos = editor.indexFromPos(cursor);
  var code = editor.getValue();

  var message = tooltipRequest(code, cursorPos);
  return tooltipObservable(channels, editor, message);
}
